kafka/kafka_source.go

```diff
@@ -98,23 +108,96 @@ func (k *KafkaSource) Generate(out chan<- interface{}) {
 	//get topics
 	kafkaTopics := []string{kconf.Topic}
 
+	var brokerList []string
 	// create consumer
-	consumer, err := consumergroup.JoinConsumerGroup(kconf.ConsumerGroupName, kafkaTopics, zookeeperNodes, config)
+	consumer, err := consumergroup.JoinConsumerGroup(kconf.ConsumerGroupName, kafkaTopics, zookeeperNodes, config, &brokerList)
 	if err != nil {
 		panic(err)
 	}
 
 	k.consumer = consumer
+
+	if k.conf.LagMonitor.Enabled {
+		//context for gracefully shutting down the offset reader goroutine
+		ctx, cancelFunc := context.WithCancel(context.Background())
+		defer cancelFunc()
+		//if polling interval is invalid then set it to default value - 5 seconds
+		if k.conf.LagMonitor.PollingInterval.Duration <= 0 {
+			k.conf.LagMonitor.PollingInterval.Duration = 5 * time.Second
+		}
+		go readProducerConsumerOffset(brokerList, kconf.Topic, k.conf.ConsumerGroupName, consumer, ctx, k.conf.LagMonitor.PollingInterval.Duration)
+	}
+
 	for message := range k.consumer.Messages() {
-		//TODO handle Create failure
+		//TODO handle Create failurex
 		kafkaMsg := k.factory.Create(message)
 
 		if k.hook != nil {
 			//TODO handle PreHook failure
 			k.hook.Pre(kafkaMsg)
 		}
+
+		//ingest sourceOffset
+		metricName := "source_offset" + "." + k.conf.ConsumerGroupName + "." + k.conf.Topic + "." + strconv.Itoa(int(kafkaMsg.GetRawMsg().Partition))
+
+		metrics.Reg.Ingest(metrics.Metric{
+			MetricType:  metrics.GAUGE,
+			MetricName:  metricName,
+			MetricValue: kafkaMsg.GetRawMsg().Offset,
```

Comment:
let's ensure we update rest of the metrics too

```suggestion
		metrics.Ingest(metrics.Metric{
			Type:  metrics.Offset,
			Name:  metricName,
			Value: kafkaMsg.GetRawMsg().Offset,
		})
```